### ISSUE TITLE
[vcpkg] Support various port upgrade strategies

### ISSUE THREAD

**Issue by yurybura** [microsoft/vcpkg#22345]
**Problem**
Certain ports have ABI guarantees depending on the port version. But vcpkg does not provide different versions of ports at the same time (with a few exceptions).

For instance Python has forward- and backwards-compatible across a minor release [1]:
> CPython’s Application Binary Interface (ABI) is forward- and backwards-compatible across a minor release (if these are compiled the same way; see Platform Considerations below). So, code compiled for Python 3.10.0 will work on 3.10.8 and vice versa, but will need to be compiled separately for 3.9.x and 3.10.x.

**Proposed solution**
vcpkg should provide different version for API incompatible versions (e.g. Python 3.7, 3.8, 3.9, etc.) in the same time and support various port upgrade strategies: "major version update" and "minor version update".
Port update examples:
* Major version update to Python 3.11 shouldn't replace exist version Python 3.10.
* Minor version update to Python 3.10.2 can replace exist version Python 3.10.1.

[1] [C API Stability](https://docs.python.org/3/c-api/stable.html)

**Comment by autoantwort** [comment]
You should be able to use [version overwrites](https://github.com/microsoft/vcpkg/blob/master/docs/examples/versioning.getting-started.md#overrides) to do that.

**Comment by yurybura** [comment]
> You should be able to use [version overwrites](https://github.com/microsoft/vcpkg/blob/master/docs/examples/versioning.getting-started.md#overrides) to do that.

Please provide me an example how to build boost-python37, boost-python38, boost-python39, boost-python310 at the same time using version overwrites.

**Comment by autoantwort** [comment]
Ah ok, but how do you use boost-python37, boost-python38, boost-python39, boost-python310 in the same project without odr violations?

**Comment by yurybura** [comment]
> Ah ok, but how do you use boost-python37, boost-python38, boost-python39, boost-python310 in the same project without odr violations? They will all have the same function names and the header names are also the same. Or what do you mean with "at the same time" when not "in the same project"?

Yes, I mean in the same port/project. For example in Boost.Python I need features python27, python37, python38, etc.

For instance I develop a Python extension (mylib.pyd) using Boost.Python. My extension users may use different Python versions. I have to build .pyd for every supported Python version (mylib37.pyd + boost-python37, mylib38.pyd + boost-python38, etc.).

**Comment by autoantwort** [comment]
But then you build your extension for python27, then for python37, then for python38, ... and not for all python versions at the same time in the same build.

**Comment by yurybura** [comment]
> But then you build your extension for python27, then for python37, then for python38, ... and not for all python versions at the same time in the same build.

Now I can build boost-python[python2,python3] at the same time. I don't need ALL versions in the same time, only major versions.

**Comment by autoantwort** [comment]
Then you have to create custom python3.9, python 3.8, ... ports like the python2 port. It is not possible to install a port twice because they would overwrite each other because they write files to the same folder.

**Comment by yurybura** [comment]
> Then you have to create custom python3.9, python 3.8, ... ports like the python2 port. It is not possible to install a port twice because they would overwrite each other because they write files to the same folder.

I cannot agree. Right now Python2 and Python3 files don't replace each other. Of course I can create custom ports, but better to have this functionality in vcpkg from the box.

**Comment by autoantwort** [comment]
>  Right now Python2 and Python3 files don't replace each other.

Because this are two independent ports.

**Comment by yurybura** [comment]
> > Right now Python2 and Python3 files don't replace each other.
> 
> Because this are two independent ports.

I cannot agree again. Python2 and Python3 are the same project even though they have independent ports in vcpkg. Having independent ports in vcpkg is a workaround for limiting the current design.

**Comment by autoantwort** [comment]
> Having independent ports in vcpkg is a workaround for limiting the current design.
True

**Comment by Hoikas** [comment]
Create your own registry with the different Python ports you are asking for.

**Comment by yurybura** [comment]
> Create your own registry with the different Python ports you are asking for.

Python is just one example. The same problem is with any library using the semantic versions (https://semver.org/).
